### ISSUE TITLE
Feature request: Allow upload_chunk_size to change in an existing installation

### ISSUE THREAD

**Issue by jmdeboer-surfsara** [filesender/filesender#1876]
**The situation as I understand it:**

In the current implementation of filesender, the upload chunk size is determined at the time of deploying the service by setting the 'upload_chunk_size' option in config.php. If this option is not specified, it defaults to 5 MB.

Once this choice has been made it is not possible to change this without making it impossible to already existing retrieve data from the storage, since the data will not reassemble correctly and decryption will fail.

**Why can this be seen this a problem?**

We are running a filesender instance, using our own S3 service as a storage backend. When the service was deployed, the default was not altered hence data is stored in 5 MB chunks (there is a suggestion in the documentation to use 40 MB for greatly increased performance but this was overlooked). 

Of course there are good reasons to use chunks. Also, one of the limits outlined in the documentation mentions PHP memory usage. But with memory being cheap, larger chunks (like 100MiB) are probably preferred.

However, not being able to change the chunk size once it has been chosen is sub-optimal imho. This leads to unnecessary load on our storage cluster resources, something we now cannot rectify.

**How might this be solved?**

I suggest making upload_chunk_size an attribute of the stored file, with the value for any new uploads being the config value as it is used now.

First of all an int field name upload_chunk_size could be added to the (file?) db table. The migration for this should also make the value of this field the same as the current upload_chunk_size for existing records.

Next, the code that retrieves the data should look at the value of this field for the file being retrieved, and use that for fetching/decrypting the chunks and reassembling the file.

Lastly, the code that does the chunking and encryption for uploads can remain largely the same; it just looks at the global upload_chunk_size, uses that, and stores this value with each file.

This will make it possible to change the chunk size on an existing installation during a maintenance window. Any data will be read with the chunk size that was in effect at the time of storage.

**A bonus:**

Using a larger chunk size (or being able to change it afterwards) comes with an additional benefit if you are not using your own S3 storage, but are using AWS; after all one of the metrics they bill you on is the number of requests. I am sure Jeff would be happy if we all use smaller chunks, that just means more requests Amazon can bill you for...

This are just some initial thoughts and I am probably overlooking things, but I am wondering what your thoughts on the matter are?

**Comment by monkeyiq** [comment]
I think this is a good idea. I would probably target the Transfer table for storing the `upload_chunk_size` used as I think it can just apply to all the files for the transfer.

There has been discussion in the past about being able to make the chunk_size dynamic. So it can ramp between a range, say 1mb to 200mb depending on your config.php. This way something starting at 4mb might move upwards if the network speed can allow it or drop back to 1mb if the network transfer is not reliable. But dynamic chunk size per file can be seen as another extension.

With that said, the crypto code uses for example
www/js/crypter/crypto_app.js:
```
crypto_chunk_size:   window.filesender.config.upload_chunk_size,
```
So it would need to be extended to pass the chunk size around instead of assuming this static default.

**Comment by WebSpider** [comment]
Yeah, we've been experimenting with the chunk sizes combined with various S3 storage vendors, and it makes a lot of difference, but also is a double-edged sword.

On the one hand, storage performance, and thus upload and download speeds for end-users really ramp up if you increase the chunk sizes. On the other hand, since a retry is per chunk, the impact of losing a chunk is also a lot bigger.

For now, in countries with high bandwidths, it seems acceptable to go up to 50MB per chunk, which is tenfold of the default.

**Comment by so010** [comment]
We from EOSC EU node FIleSender project want to second that feature request. We want to finetune the upload_chunk_size but are locked in since there are always active transfers which will be lost by changing that parameter.

**Comment by adminnr1** [comment]
This is a feature request that would benefit our installation greatly as well. Please do develop!

**Comment by WebSpider** [comment]
To give an estimate of the improvements we've seen: 
- When changing the chunk size from 5MB to 40MB, the performance in mbit/s on the same backend storage we have been testing with (S3 via ceph) increased x10. 
- When changing it to 80MB, performance went up linearly.
- This was on 3.0rc3

Important detail: We used ramfs for all locations that tempfiles were stored in this test.

**Comment by madsi1m** [comment]
Also in favour of this as long as it also is compatible with the chunked filesystem class.

**Comment by monkeyiq** [comment]
> Also in favour of this as long as it also is compatible with the chunked filesystem class.

The PR with these changes would have to test that explicitly. When I added some crypto options in the past I also added test cases to the selenium to make sure old and new transfers could be still used. So with this having the config change (with chunked files) and then have the test suite download previous uploads and check hashes would be the only way to add this and be confident that folks will not have headaches because of it.

**Comment by adminnr1** [comment]
Great to see that there is so much support for this feature request. Is there an ETA for this feature @monkeyiq ?

**Comment by monkeyiq** [comment]
I am hoping to have it out in the next month or two. It is probably best to set a conservative time line and beat it.

It is high on my radar now and I have been working on it.

**Comment by monkeyiq** [comment]
For those who are interested in this feature I have a PR at https://github.com/filesender/filesender/pull/2166 which should allow this. The PR mentions the testing I have done on it so far.

I welcome feedback and hope to release this PR in the next 2.x series release.

**Comment by monkeyiq** [comment]
@madsi1m As with all of my current work flow this will be ported into 3.x with an upcoming merge. This will likely be done once it is working and released in 2.x. I have specifically targeted the two built in disk based storage engines in the new Acceptance test for the PR 
https://github.com/monkeyiq/filesender/blob/fb36348be853ac861d396cd270426e742293563c/unittests/codeception/acceptance/ConfigurationABCest.php#L145

**Comment by madsi1m** [comment]
@monkeyiq happy to do testing if you have a test v3 branch i can pull from

**Comment by madsi1m** [comment]
made my own, https://github.com/madsi1m/filesender/tree/dev3-chunks

**Comment by madsi1m** [comment]
Seems to work on v3 with chunked filesystem, interestingly our previously tuned values for our old storage were spot on for our totally different current storage solution.

What made a bigger difference was re-tuning terasender_workers to a larger number.

**Comment by monkeyiq** [comment]
I also had a go at a v3 port, which is at https://github.com/monkeyiq/filesender/tree/2025/march/chunk-size-vary-d3

It is probably a good idea for me to diff mine and yours to see if you discovered some other items that needed updating :)

I think the chunk size variation has the biggest impact on folks using the cloud storage back end.

**Comment by monkeyiq** [comment]
I found that `terasender_workers` would plateau at a certain number, IIRC somewhere in the 8 to 12 range. I want to get back to investigating that, if it is a browser limitation, web server limitation, or where the choke point is.

**Comment by madsi1m** [comment]
I think browsers allow more now, as i also remember it being about 8ish, now i can do 20 without problem after that it plateaus and even more (after disabling the config checks just to test) it actually degrades. I did some research , there doesn't seem to be a thread limit anymore, it seems to be a memory limit for threads so as long as it doesn't hog memory you should be fine.
